substrate/runtime/timestamp/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Substrate Demo.
+
+// Substrate Demo is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate Demo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Timestamp manager: just handles the current timestamp.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg_attr(test, macro_use)] extern crate substrate_runtime_std as rstd;
+#[macro_use] extern crate substrate_runtime_support as runtime_support;
+#[cfg(test)] extern crate substrate_runtime_io as runtime_io;
+extern crate substrate_codec as codec;
+
+#[cfg(feature = "std")] #[macro_use] extern crate serde_derive;
+#[cfg(feature = "std")] extern crate serde;
+
+use runtime_support::storage::StorageValue;
+use runtime_support::PublicPass;
+
+pub trait Trait {
+	type Timestamp: codec::Slicable;
+	type PublicAux;
+	type PrivAux;
+}
+pub struct Module<T: Trait>(::std::marker::PhantomData<T>);
+
+// TODO: create from storage_items macro.
+/*
+storage_items! {
+	trait Trait;
+	pub Now: b"tim:val" => required Timestamp;
+}
+*/
+pub struct Now<T: Trait>(::std::marker::PhantomData<T>);
+impl<T: Trait> ::runtime_support::storage::generator::StorageValue<T::Timestamp> for Now<T> {
+	type Query = T::Timestamp;
+
+	/// Get the storage key.
+	fn key() -> &'static [u8] {
+		b"tim:val"
+	}
+
+	/// Load the value from the provided storage instance.
+	fn get<S: ::runtime_support::GenericStorage>(storage: &S) -> Self::Query {
+		storage.require(<Self as ::runtime_support::storage::generator::StorageValue<T::Timestamp>>::key())
+	}
+
+	/// Take a value from storage, removing it afterwards.
+	fn take<S: ::runtime_support::GenericStorage>(storage: &S) -> Self::Query {
+		storage.take_or_panic(<Self as ::runtime_support::storage::generator::StorageValue<T::Timestamp>>::key())
+	}
+}
+// TODO: all `Storage*` functions should take `self` parameter to avoid need for obscure `<Now<TraitImpl>>::`
+// syntax. this can be cleaned up once we get `type`s in `impl`s.
+/*impl<T: Trait> Module<T> {
+	pub const NOW: Now<T> = Now(::std::marker::PhantomData::<T>);
+}*/
+
+/*
+// This would be nice but not currently supported.
+impl<T: Trait> Module<T> {
+	type Now = super::Now<T>;
+}*/
+
+pub trait StorageItems {
+	type NowS;
+}
+impl<T: Trait> StorageItems for Module<T> {
+	type NowS = Now<T>;
+}
+
+impl<T: Trait> Module<T> {
+	pub fn get() -> T::Timestamp { <Now<T>>::get() }
```

Comment:
and if the timestamp were the only generic param here (and we went a little more general than "some serializable type" to get to "some timestamp which can be get and set") the implementation here would still be the same, while still extending to timestamps which are out of storage _and_ being testable without setting up storage context